Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp

```diff
@@ -832,9 +833,15 @@ void GraphicsContextCG::fillRect(const FloatRect& rect, Gradient& gradient, cons
         gradient.paint(layerContext);
         CGContextDrawLayerInRect(context, rect, layer.get());
     } else {
-        CGContextClipToRect(context, rect);
+        bool skipClip = getCTM().mapRect(rect).contains(CGContextGetDeviceClipBoundingBox(context));
+        if (!skipClip) {
+            stateSaver.save();
+            CGContextClipToRect(context, rect);
+        }
         CGContextConcatCTM(context, gradientSpaceTransform);
         gradient.paint(*this);
+        if (skipClip)
+            CGContextConcatCTM(context, CGAffineTransformInvert(gradientSpaceTransform));
```

Comment:
Are you also trying to save the cost of a save/restore in the no-clip case? The commit message doesn't mention that either.

Reply:
Indeed those two should be equivalent, I will try that instead